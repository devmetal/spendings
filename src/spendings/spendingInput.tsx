import React, { ChangeEvent, FC, useState, useEffect } from 'react';
import FormControl from 'react-bootstrap/FormControl';
import InputGroup from 'react-bootstrap/InputGroup';
import Button from 'react-bootstrap/Button';

export interface SpendingInputProps {
  id?: number;
  amount?: number;
  name?: string;
  disabled: boolean;
  onCancel?: () => void;
  onSave: (spending: { id?: number; amount: number; name: string }) => void;
}

export const SpendingInput: FC<SpendingInputProps> = ({
  id,
  amount,
  name,
  onSave,
  disabled,
}) => {
  const [amountVal, setAmountVal] = useState<string>(() => `${amount ?? ''}`);
  const [nameVal, setNameVal] = useState<string>(() => `${name ?? ''}`);

  useEffect(() => {
    setAmountVal(`${amount ?? ''}`);
  }, [amount]);

  useEffect(() => {
    setNameVal(`${name ?? ''}`);
  }, [name]);

  const handleSave = () => {
    if (!nameVal) {
      return;
    }

    if (!amountVal) {
      return;
    }

    const amountNum = Number(amountVal);
    if (amountNum < 1) {
      return;
    }

    onSave({
      id,
      name: nameVal,
      amount: Number(amountVal),
    });
  };

  const handleNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    setNameVal(e.target.value);
  };

  const handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
    setAmountVal(e.target.value);
  };

  return (
    <InputGroup className="mb-3">
      <InputGroup.Prepend>
        <InputGroup.Text>Name</InputGroup.Text>
      </InputGroup.Prepend>
      <FormControl
        data-testid="name"
        disabled={disabled}
        type="text"
        value={nameVal}
        onChange={handleNameChange}
      />
      <InputGroup.Prepend>
        <InputGroup.Text>Amount</InputGroup.Text>
      </InputGroup.Prepend>
      <FormControl
        data-testid="amount"
        disabled={disabled}
        type="number"
        min="1"
        value={amountVal}
        onChange={handleAmountChange}
      />
      <InputGroup.Prepend>
        <Button
          disabled={disabled}
          onClick={handleSave}
          variant="outline-primary"
        >
          Save
        </Button>
      </InputGroup.Prepend>
    </InputGroup>
  );
};
