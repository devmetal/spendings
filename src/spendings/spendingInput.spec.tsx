import React from 'react';
import { fireEvent, render, RenderResult } from '@testing-library/react';
import { SpendingInput } from './spendingInput';

test('Sanity snapshot test', () => {
  const { container } = render(
    <SpendingInput disabled={false} onSave={jest.fn()} />
  );
  expect(container).toMatchSnapshot();
});

test('Pre filles values', () => {
  const { container } = render(
    <SpendingInput
      disabled={false}
      id={1}
      amount={500}
      name="Hello"
      onSave={jest.fn()}
    />
  );
  expect(container).toMatchSnapshot();
});

describe('onSave callback', () => {
  let amount: HTMLElement;
  let name: HTMLElement;
  let saveBtn: HTMLElement;
  let onSave: () => {};

  beforeEach(() => {
    onSave = jest.fn();

    const { getByTestId, getByText } = render(
      <SpendingInput onSave={onSave} disabled={false} />
    );

    amount = getByTestId('amount');
    name = getByTestId('name');
    saveBtn = getByText('Save');
  });

  it('wont fires when name is empty', () => {
    fireEvent.change(amount, { target: { value: '100' } });
    fireEvent.click(saveBtn);
    expect(onSave).not.toBeCalled();
  });

  it('wont fires when amount is empty', () => {
    fireEvent.change(name, { target: { value: 'valami' } });
    fireEvent.click(saveBtn);
    expect(onSave).not.toBeCalled();
  });

  it('wont fires when amount is invalid', () => {
    fireEvent.change(amount, { target: { value: '-100' } });
    fireEvent.change(name, 'valami');
    fireEvent.click(saveBtn);
    expect(onSave).not.toBeCalled();
  });

  it('fires when amount and name filled and valid', () => {
    fireEvent.change(amount, { target: { value: '100' } });
    fireEvent.change(name, { target: { value: 'valami' } });
    fireEvent.click(saveBtn);
    expect(onSave).toBeCalledWith({ name: 'valami', amount: 100 });
  });
});
