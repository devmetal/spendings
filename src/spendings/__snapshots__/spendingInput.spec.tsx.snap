// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pre filles values 1`] = `
<div>
  <div
    class="mb-3 input-group"
  >
    <div
      class="input-group-prepend"
    >
      <span
        class="input-group-text"
      >
        Name
      </span>
    </div>
    <input
      class="form-control"
      data-testid="name"
      type="text"
      value="Hello"
    />
    <div
      class="input-group-prepend"
    >
      <span
        class="input-group-text"
      >
        Amount
      </span>
    </div>
    <input
      class="form-control"
      data-testid="amount"
      min="1"
      type="number"
      value="500"
    />
    <div
      class="input-group-prepend"
    >
      <button
        class="btn btn-outline-primary"
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;

exports[`Sanity snapshot test 1`] = `
<div>
  <div
    class="mb-3 input-group"
  >
    <div
      class="input-group-prepend"
    >
      <span
        class="input-group-text"
      >
        Name
      </span>
    </div>
    <input
      class="form-control"
      data-testid="name"
      type="text"
      value=""
    />
    <div
      class="input-group-prepend"
    >
      <span
        class="input-group-text"
      >
        Amount
      </span>
    </div>
    <input
      class="form-control"
      data-testid="amount"
      min="1"
      type="number"
      value=""
    />
    <div
      class="input-group-prepend"
    >
      <button
        class="btn btn-outline-primary"
        type="button"
      >
        Save
      </button>
    </div>
  </div>
</div>
`;
